vernac/classes.ml

```diff
@@ -189,163 +337,11 @@ let new_instance ?(abstract=false) ?(global=false) ?(refine= !refine_instance)
-      let term = Option.map (to_constr ~abort_on_undefined_evars:false sigma) term in
-        if not (Evd.has_undefined sigma) && not (Option.is_empty term) then
-	  declare_instance_constant k pri global imps ?hook id decl
-            poly sigma (Option.get term) termtype
-        else if program_mode || refine || Option.is_empty term then begin
-	  let kind = Decl_kinds.Global, poly, Decl_kinds.DefinitionBody Decl_kinds.Instance in
-            if program_mode then
-	      let hook vis gr _ =
-		let cst = match gr with ConstRef kn -> kn | _ -> assert false in
-                  Impargs.declare_manual_implicits false gr ~enriching:false [imps];
-                  let pri = intern_info pri in
-		  Typeclasses.declare_instance (Some pri) (not global) (ConstRef cst)
-	      in
-	      let obls, constr, typ =
-		match term with 
-		| Some t -> 
-		  let obls, _, constr, typ = 
-                    Obligations.eterm_obligations env id sigma 0 t termtype
-		  in obls, Some constr, typ
-		| None -> [||], None, termtype
-	      in
-              let hook = Lemmas.mk_hook hook in
-              let ctx = Evd.evar_universe_context sigma in
-		ignore (Obligations.add_definition id ?term:constr
- 			~univdecl:decl typ ctx ~kind:(Global,poly,Instance) ~hook obls);
-		id
-	    else
-	      (Flags.silently 
-	       (fun () ->
-                  (* spiwack: it is hard to reorder the actions to do
-                     the pretyping after the proof has opened. As a
-                     consequence, we use the low-level primitives to code
-                     the refinement manually.*)
-                let gls = List.rev (Evd.future_goals sigma) in
-                let sigma = Evd.reset_future_goals sigma in
-                Lemmas.start_proof id ~pl:decl kind sigma (EConstr.of_constr termtype)
-		(Lemmas.mk_hook
-                  (fun _ -> instance_hook k pri global imps ?hook));
-                 (* spiwack: I don't know what to do with the status here. *)
-		if not (Option.is_empty term) then
-                  let init_refine =
-                    Tacticals.New.tclTHENLIST [
-                      Refine.refine ~typecheck:false (fun sigma -> (sigma,EConstr.of_constr (Option.get term)));
-                      Proofview.Unsafe.tclNEWGOALS (CList.map Proofview.with_empty_state gls);
-                      Tactics.New.reduce_after_refine;
-                    ]
-                  in
-		  ignore (Pfedit.by init_refine)
-		else if Flags.is_auto_intros () then
-		  ignore (Pfedit.by (Tacticals.New.tclDO len Tactics.intro));
-		(match tac with Some tac -> ignore (Pfedit.by tac) | None -> ())) ();
-	       id)
-	end
-      else CErrors.user_err Pp.(str "Unsolved obligations remaining."))
-	
+  if abstract then
+    do_abstract_instance env sigma ?hook ~global ~poly k u ctx ctx' pri decl imps subst id
+  else
+    do_transparent_instance env env' sigma ?hook ~refine ~tac ~global ~poly ~program_mode cty k u ctx ctx' pri decl imps subst id props len
```

Comment:
Looks a good improvement! May I suggest to also wrap around lines like this one, while you are at it?

Reply:
Sorry @gares , what do you mean by wrap?

Reply:
Ahhh you mean 80 char limit. Sure, I didn't do it as this is not enforced in Coq, but I'll give it a try.

Reply:
I like 80 chars, even if I don't mind enforcing it with a linter, since it would make sed-generated-patches harder to write, and we need many of these to clean up the sources.

But this line is very very long, and it is new, so I hope you take the occasion to break it ;-)

Reply:
> But this line is very very long, and it is new, so I hope you take the occasion to break it ;-)

Sure, I will just use what `ocamlfmt / ocp-indent` do.

Reply:
Anyways the real problem here is not indentation, but lack of structure/abstraction.

Reply:
Thanks! (btw, also the declaration of the function is very long...)

Reply:
I get some weird formatting for the declaration, I think tabs are confusing the indent tool.